### PR TITLE
Add the ability to change module name

### PR DESCRIPTION
**Pull request by AlexeiVainshtein**

### CODE REVIEW

**Comment by eyalbe4** [comment]
Great work @AlexeiVainshtein!
See my inline suggestions.

File: artifactory/commands/generic/download.go [hunk 1]
```diff
@@ -113,6 +113,9 @@ func (dc *DownloadCommand) Run() error {
 		buildDependencies := convertFileInfoToBuildDependencies(filesInfo)
 		populateFunc := func(partial *buildinfo.Partial) {
 			partial.Dependencies = buildDependencies
+			if dc.buildConfiguration.Module != "" {
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply artifactory/commands/generic/download.go]: Done

File: artifactory/commands/generic/upload.go [hunk 2]
```diff
@@ -131,6 +131,9 @@ func (uc *UploadCommand) upload() error {
 		buildArtifacts := convertFileInfoToBuildArtifacts(filesInfo)
 		populateFunc := func(partial *buildinfo.Partial) {
 			partial.Artifacts = buildArtifacts
+			if uc.buildConfiguration.Module != "" {
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply artifactory/commands/generic/upload.go]: Done

File: artifactory/cli.go [hunk 3]
```diff
@@ -691,19 +679,32 @@ func getDownloadFlags() []cli.Flag {
 	}...)
 }
 
-func getBuildToolFlags() []cli.Flag {
+func getBuildToolFlags(buildNameMsg, buildNumberMsg string) []cli.Flag {
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply artifactory/cli.go]: Done

File: artifactory/cli.go [hunk 4]
```diff
@@ -691,19 +679,32 @@ func getDownloadFlags() []cli.Flag {
...
-			Usage: "[Optional] Providing this option will collect and record build info for this build number. If you provide a build name (using the --build-name option) and do not provide a build number, a build number will be automatically generated.` `",
+			Usage: buildNumberMsg,
 		},
 	}
 }
 
+func getBuildToolAndModuleFlags(buildNameMsg, buildNumberMsg string) []cli.Flag {
+	return append(getBuildToolFlags(buildNameMsg, buildNumberMsg), cli.StringFlag{
+		Name:  "module",
+		Usage: "[Optional] Providing this option will collect and record the artifacts under this module name for this build number.` `",
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply artifactory/cli.go]: Changed it Optional module name for the build-info

File: artifactory/cli.go [hunk 5]
```diff
@@ -2407,9 +2405,9 @@ func validateCommonContext(c *cli.Context) {
 	}
 }
 
-func validateBuildParams(buildName, buildNumber string) {
-	if (buildName == "" && buildNumber != "") || (buildName != "" && buildNumber == "") {
-		cliutils.ExitOnErr(errors.New("The build-name and build-number options cannot be sent separately."))
+func validateBuildParams(buildName, buildNumber, module string) {
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply artifactory/cli.go]: Yes, I've changed it.

File: buildinfo_test.go [hunk 6]
```diff
@@ -416,6 +413,75 @@ func uploadFilesAndGetBuildInfo(t *testing.T, buildName, buildNumber, buildUrl s
...
+	// Add the build number argument to the download command
+	downloadArgs = append(downloadArgs, args[2])
+	downloadFiles(t, downloadArgs...)
+}
+
+func TestModuleName(t *testing.T) {
+	initArtifactoryTest(t)
+	buildName := "cli-test-build"
+	tests := []struct {
+		name                 string
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply buildinfo_test.go]: No. It's the test name. I've changed it to testName

File: buildinfo_test.go [hunk 7]
```diff
@@ -416,6 +413,75 @@ func uploadFilesAndGetBuildInfo(t *testing.T, buildName, buildNumber, buildUrl s
...
+		t.Error(err)
+	}
+	args = append(args, "--spec="+specFile)
+	artifactoryCli.Exec(args...)
+}
+
+// The first argument in the args slice is for upload,
+// The second argument is for download command
+// The third is build number argument.
+func uploadAndDownload(t *testing.T, args ...string) {
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply buildinfo_test.go]: I've removed this function. I've created a different struct for a command (function upload or download and the args for that function) and a slice that holds those commands. Running if a loop to trigger it all.

File: buildtools_test.go [hunk 8]
```diff
@@ -409,17 +423,46 @@ func runGo(t *testing.T, wd, gopath, project1Path string, copyDirs bool) {
...
 	} else {
-		validateBuildInfo(buildInfo, t, 8, 0)
+		validateBuildInfo(buildInfo, t, 8, 0, module)
 	}
 
-	err = os.Chdir(wd)
+	inttestutils.DeleteBuild(artifactoryDetails.Url, buildName, artHttpDetails)
+}
+
+func prepareProject(configDestDir string, t *testing.T, copyDirs bool) {
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply buildtools_test.go]: Done

**Pull request merged**